Mixed chart

loicbourgois opened the pull request:
As mentioned in #4184, #4587, #4096, the mixed chart functionality is a bit buggy : it only allows `line` charts to be added to a `bar` chart.

### Original comment
I found a way to display `line`, `bar`, `scatter` on a chart of type `line` or `bar`.
I'm afraid it doesn't address the root problem, but it's a net improvement.
You can find a working example here :  https://codepen.io/anon/pen/zpWMOj

### More mixed chart
I made changes to `initConfig` in the core controller, to merge options from the child charts in the parent chart.
Supported parents are : `line`, `bar`, `undefined`
Supported childs are : `scatter`, `line`, `bubble`, `bar`
Here is an example with the changes until commit https://github.com/chartjs/Chart.js/pull/5134/commits/042920e7845fa9908416135994e155d1b6007f8d : [mixed](https://jsfiddle.net/dm4unaog/5/)
You can compare it to the current behavior : [2.7.1](https://jsfiddle.net/mjnrr9g3/5/)

simonbrunel commented on package-lock.json:
```diff
@@ -0,0 +1,14260 @@
+{
```
What are the motivations behind adding this file to the repo? Does it fix anything? Does it enhance anything? If not, let's move forward and keep that file outside the repository and avoid another PR that wastes time for everyone.

loicbourgois replied:
According to the [npm doc](https://docs.npmjs.com/files/package-lock.json) :
>package-lock.json is automatically generated for any operations where npm modifies either the node_modules tree, or package.json. It describes the exact tree that was generated, such that subsequent installs are able to generate identical trees, regardless of intermediate dependency updates.
>This file is intended to be committed into source repositories, and serves various purposes:
>- Describe a single representation of a dependency tree such that teammates, deployments, and continuous integration are guaranteed to install exactly the same dependencies.
>- Provide a facility for users to "time-travel" to previous states of node_modules without having to commit the directory itself.
>- To facilitate greater visibility of tree changes through readable source control diffs.
>- And optimize the installation process by allowing npm to skip repeated metadata resolutions for previously-installed packages.

If this file is not to be added, then `.gitignore` should probably be modified accordingly.

simonbrunel replied:
I was pretty sure we added that file to `.gitignore`! ([PR submitted](/chartjs/Chart.js/pull/5138))

I know about that npm description, but after some investigations, internet reads and debates, it doesn't seem to bring any benefit to our project, only potential maintaining issues. We already spent lot of time discussing that point and until now we didn't experience any issue not having that file on the repo.

etimberg commented on src/controllers/controller.scatter.js:
```diff
@@ -37,6 +37,10 @@ defaults._set('scatter', {
 
 	// Scatter charts use line controllers
-	Chart.controllers.scatter = Chart.controllers.line;
+	Chart.controllers.scatter = Chart.controllers.line.extend({
```
What's the reason for overriding and returning false here? Shouldn't that have already happened?

loicbourgois replied:
From what i understand, with mixed chart, the default style applied is the style of the 'main' chart.
And for now, only `bar` or `line` are valid main types.
So the option `showLines` is never set to `false`. You'd have to do it manually in the child chart.
You can try commenting the line n°8187 and n°19040 in here https://codepen.io/anon/pen/zpWMOj
And you can check the console for `Chart.config.data.datasets[0].showLine`.

loicbourgois replied:
But now that I think about it, is being able to show lines in a scatter chart a wanted feature ?  
If so, then this overriding would break it.

etimberg replied:
I think lines on a scatter chart is something that should be possible and its currently the default behaviour

loicbourgois replied:
So does that mean a test should be added via another PR to make sure it stays that way in the future ?

etimberg replied:
I would think adding a test to ensure that a scatter chart can optionally use a line or not is a good idea. @simonbrunel thoughts?

simonbrunel replied:
I don't see why scatter charts couldn't display lines, so yes these `lineEnabled` changes should be reverted. We could add a test indeed, but in a different PR.

etimberg commented on src/controllers/controller.bar.js:
```diff
@@ -284,7 +285,14 @@ module.exports = function(Chart) {
-			return this.getScaleForId(this.getIndexScaleId());
+			var scale = this.getScaleForId(this.getIndexScaleId());
+			if (scale.options.categoryPercentage === undefined) {
+				scale.options.categoryPercentage = defaultBar.scales.xAxes[0].categoryPercentage;
```
[nit] I would put `defaultBar.scales.xAxes[0]` into a local variable to save a few bytes

benmccann commented on samples/charts/mixed-chart.html:
```diff
@@ -0,0 +1,122 @@
+				childs.forEach(function(child) {
+					c++;
+					var data = [];
+					for(var i = 0 ; i < size ; i++) {
```
should update all the `if` and `for` statements in this PR to include a space before `(` for consistency with the rest of the Chart.js codebase

benmccann commented on src/core/core.controller.js:
```diff
@@ -31,10 +31,74 @@ module.exports = function(Chart) {
+				}
+			});
+			if (isMixed) {
+				if (hasBarType) { // [3]
```
you could save a lot of lines by removing the `if/else` and changing the second parameter of the `configMerge` to be `hasBarType ? defaults.bar : defaults[config.type || 'line']`

loicbourgois replied:
Good idea, thanks!

benmccann commented on src/core/core.controller.js:
```diff
@@ -31,10 +31,74 @@ module.exports = function(Chart) {
+
+		// Merge options for mixed charts
+		// [1] Does not work if axes are stacked
+		// [2] Only triggered if there are at least 2 different types
```
it seems like the code would be simpler if you always did it. is this required?

loicbourgois replied:
If I just replace `if (isMixed) {` with `if (true/*isMixed*/) {`, `gulp test` gives me the following : 
```
Chrome 63.0.3239 (Linux 0.0.0): Executed 683 of 687 (130 FAILED) (skipped 4) ERROR (8.176 secs / 7.324 secs)
Firefox 52.0.0 (Linux 0.0.0): Executed 683 of 687 (130 FAILED) (skipped 4) (14.328 secs / 12.872 secs)
```
:/

And for stacked axes, without the check, 2 tests are failing.

loicbourgois replied:
Fixed it ! 
```
if (dataset.type !== undefined) {
	config.options = helpers.configMerge(defaults[dataset.type], config.options || {});
}
```

benmccann commented on src/core/core.controller.js:
```diff
@@ -31,10 +31,74 @@ module.exports = function(Chart) {
+		//     we set the main chart to 'line'
+		//     This way, users can use a mixed chart without having to set a main type
+		var isStacked = false;
+		var optionstmp = helpers.configMerge(
```
I probably would call this `optionsTmp`

benmccann commented on src/core/core.controller.js:
```diff
@@ -31,10 +31,31 @@ module.exports = function(Chart) {
 
+
+		// Merge options for mixed charts
+		// If there is at least one 'bar' chart, main type is set to 'bar'
```
Can you expand the comment to explain why `If there is at least one 'bar' chart, main type is set to 'bar'`

One thing I'm wondering about is if there's a more generic way to do this. I'm worried that this won't work if you want to mix a financial chart and line chart for example (https://github.com/chartjs/chartjs-chart-financial)

simonbrunel replied:
I share the same concerns: wouldn't be better to add support for bar specific options (scale) at the dataset level and avoid touching the options logic? Actually, I think that can cause breaking changes, especially [the logic that merges all dataset defaults on top of each others](/chartjs/Chart.js/pull/5134/files#diff-de7249441169d0e646c53d8220e8b0c5R55). I'm not even sure it works great since the last dataset (options) overwrites the previous ones.

For example, with the following types: `line` > `scatter`, the value of `showLines` will be `false` but `scatter` > `line` generates `showLines: true` (not consistent). Same for `line` > `bubble`, the tooltip options will break for the line.

benmccann replied:
I agree that we should try to have all options supported at the dataset level

simonbrunel commented:
I don't know exactly what are the "blocking" issues that make mixed charts impossible, but unfortunately, charts are not consistent in their default options and I don't think there is a magic way to generate the "good" defaults for any mixed charts (e.g. can't merge `legendCallback`, `tooltips.callbacks.label`, etc.).

IMO, if no global type is given, then only the global defaults should be merged and the `dataset.type` should become mandatory. The user would have to provide the correct options that works great with his datasets and on our side we should make sure that we can handle mixed dataset types.

benmccann commented:
The are two main issues with mixed charts.

The first is that we have scatter and line charts which are the same exact thing, but with different default values of `showLines`. It's problematic that we try to set this variable both via chart type and via a config option.

The second is that `categoryPercentage` and `barPercentage` do not have default values unless the overall chart type is `bar`. That means that you cannot have a `line` chart with `bar` dataset because those values will be undefined. It currently works only if you swap them (i.e. `bar` chart with `line` dataset)

simonbrunel commented:
Are these really blocking issues?
- `categoryPercentage` and `barPercentage` can be [explicitly defined](https://jsfiddle.net/simonbrunel/vkw1x4u3/2/) or set as [scale defaults](https://jsfiddle.net/simonbrunel/vkw1x4u3/1/)
- `showLines` is [supported per dataset as `showLine`](/chartjs/Chart.js/blob/7dc71d009ba13617bdde1e5c3e59b9c47dc34fe8/src/controllers/controller.line.js#L30) ([fiddle](https://jsfiddle.net/simonbrunel/7bxakwhm/1/))

With 2.7.1, mixed line + bar + scatter + bubble seems to work ([fiddle](https://jsfiddle.net/simonbrunel/z87e0LbL/)). Though, it only works if the chart type is `line` or `bar`, that would need to be investigated. But it doesn't seem better with the suggested changes?

benmccann commented:
I've explicitly defined `categoryPercentage` and `barPercentage` in the past when I want to use a mixed chart. However, it took me half an hour of debugging to realized I needed to do this. Hopefully we can make it easier for users. I would suggest that we move the default values for these config options from `controller.bar.js` to `core.scale.js`

loicbourgois commented:
I updated the merging of options, removed the changes to scatter and line controllers, and added a new test.
The test however doesn't seem to fail. I'm using `expect(meta.data[0].draw.calls.count()).toBe(1);`
What would be the appropriate way to verify that all the charts are actually displaying ? 
With these latest commits, here are fiddles to compare it to master : 
[Master + parent defined](https://jsfiddle.net/loicbourgois/dg3vxgee/) : 
[Mixed + parent defined](https://jsfiddle.net/loicbourgois/a3xsxsqv/)
[Mixed + parent undefined](https://jsfiddle.net/loicbourgois/d422bt23/)

benmccann commented:
I tend to agree with @simonbrunel that the way to fix these issues isn't by merging options. I think it becomes really difficult to understand the code to know which order options get applied, which override which, etc.

I think we should basically try to get rid of chart type options and have dataset and scale options instead. E.g. see my last two comments for an approach with bar charts. And here's a PR to fix the line / scatter issue: https://github.com/chartjs/Chart.js/pull/5151

benmccann commented:
I talked to Simon on Slack today about this. He suggested the right long term solution is the one outlined in https://github.com/chartjs/Chart.js/issues/5191 and that he will hopefully try to see if it's feasible if he has time

loicbourgois commented:
It sounds like a better solution indeed. 
Will this allow to have an `undefined` main type ?

simonbrunel commented:
Thanks @loicbourgois! I'm closing this PR because I think #5999 will help handling mixed chart in a better way than the `typesByPriotity` approach suggested here which doesn't seem a scalable, stable and durable solution.

nazariyv commented:
Can the scatter plot take the x-values that are different to the line chart, but are on the same scale as the line chart. i.e. assume you have x-axes range for the line chart to be `[0.5, 1.0, 2.0, 2.5, 3.0]` and for the scatter chart, one of the points takes an x value of 0.65. As far as I can tell, it gets moved to 0.5...

benmccann commented:
@nazariyv let's keep the discussion of that question on https://github.com/chartjs/Chart.js/issues/6409. thanks!
